contracts/contract_call.go

```diff
@@ -0,0 +1,120 @@
+package contracts
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/accounts/abi"
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/common/hexutil"
+	"github.com/celo-org/celo-blockchain/core/vm"
+	"github.com/celo-org/celo-blockchain/log"
+)
+
+// Method represents a contract's method
+type Method struct {
+	abi    *abi.ABI
+	method string
+	maxGas uint64
+}
+
+// NewMethod creates a new contract message
+func NewMethod(abi *abi.ABI, method string, maxGas uint64) Method {
+	return Method{
+		abi:    abi,
+		method: method,
+		maxGas: maxGas,
+	}
+}
+
+func (am Method) Bind(contractAddress common.Address) *BoundMethod {
+	return &BoundMethod{
+		Method:         am,
+		resolveAddress: noopResolver(contractAddress),
+	}
+}
+
+// encodeCall will encodes the msg into []byte format for EVM consumption
+func (am Method) encodeCall(args ...interface{}) ([]byte, error) {
+	return am.abi.Pack(am.method, args...)
+}
+
+// decodeResult will decode the result of msg execution into the result parameter
+func (am Method) decodeResult(result interface{}, output []byte) error {
+	if result == nil {
+		return nil
+	}
+	return am.abi.Unpack(result, am.method, output)
+}
+
+func NewBoundMethod(contractAddress common.Address, abi *abi.ABI, methodName string, maxGas uint64) *BoundMethod {
+	return NewMethod(abi, methodName, maxGas).Bind(contractAddress)
+}
+
+func NewRegistryContractMethod(registryId common.Hash, abi *abi.ABI, methodName string, maxGas uint64) *BoundMethod {
+	return &BoundMethod{
+		Method: NewMethod(abi, methodName, maxGas),
+		resolveAddress: func(caller vm.SystemEVM) (common.Address, error) {
+			return resolveAddressForCall(caller, registryId, methodName)
+		},
+	}
+}
+
+type BoundMethod struct {
+	Method
+	resolveAddress func(vm.SystemEVM) (common.Address, error)
+}
+
+func (bm *BoundMethod) VMQuery(caller vm.SystemEVM, result interface{}, args ...interface{}) error {
+	return bm.Query(caller, result, VMAddress, args...)
+}
+
+func (bm *BoundMethod) Query(caller vm.SystemEVM, result interface{}, sender common.Address, args ...interface{}) error {
+	return bm.run(caller, result, true, sender, nil, args...)
+}
+
+func (bm *BoundMethod) VMExecute(caller vm.SystemEVM, result interface{}, value *big.Int, args ...interface{}) error {
```

Comment:
There's a lack of symmetry here: `VMQuery` calls `Query` but `VMExecute` calls `run` directly.  I don't think there's a big difference between the two, you I might lean a little towards calling `.Execute()`.  But either way it'd be nicer if they were consistent.

Reply:
Remove the `VM..` variants.